### DISCUSSION TITLE
[BUG] ERROR: from element /GstPipeline:pipeline0/GstV4l2Src:v4l2src0: Internal data stream error.

### DISCUSSION THREAD

**Discussion by Karthiktmv** [awslabs/amazon-kinesis-video-streams-producer-sdk-cpp#980] [Q&A] [answered]
**Logging**
ERROR: from element /GstPipeline:pipeline0/GstV4l2Src:v4l2src0: Internal data stream error.
Additional debug info:
../libs/gst/base/gstbasesrc.c(3127): gst_base_src_loop (): /GstPipeline:pipeline0/GstV4l2Src:v4l2src0:
streaming stopped, reason not-negotiated (-4)

**Describe the bug**
I have Lenovo Legion Y540 laptop & I am trying to set up AWS Kinesis Video stream with gstreamer in my oracle virtual box with ubuntu 22.04 OS.
I have followed all the steps and the whole setup is complete without any problem. 

I want to stream video from my laptop's integrated camera to AWS KVS via this cpp sdk. 
I can use my camera properly inside my ubuntu in virtualbox with "cheese ." 

I used this command from https://github.com/awsdocs/AWS-Kinesis-Video-Documentation/blob/master/doc_source/examples-gstreamer-plugin.md:-

Example 2: Encode and Stream Video from a USB Camera on Ubuntu:-

$ gst-launch-1.0 v4l2src do-timestamp=TRUE device=/dev/video0 ! videoconvert ! video/x-raw,format=I420,width=640,height=480,framerate=30/1 ! x264enc  bframes=0 key-int-max=45 bitrate=500 ! video/x-h264,stream-format=avc,alignment=au,profile=baseline ! kvssink stream-name="YourStreamName" storage-size=512 access-key="YourAccessKey" secret-key="YourSecretKey" aws-region="YourAWSRegion"

After using  v4l2-ctl --list-devices, my devices also show in output, but i am getting this error:-

![ss1](https://user-images.githubusercontent.com/111423487/197115049-01000b36-8908-4c2b-b789-1217f3b0e42b.png)

![ss2](https://user-images.githubusercontent.com/111423487/197115163-acdb036a-8513-46be-aa70-1fa8ac2cef61.png)

Here is an output of gst-device-monitor-1.0:-

![ss3](https://user-images.githubusercontent.com/111423487/197116207-3a98297e-6e47-42a4-98c6-d54ca6235d26.png)

Please help me in resolving this issue.

OS:- Windows 11 with Oracle VirtualBox 6.1.40 with Ubuntu 22.04

**Comment by disa6302** [comment, marked as answer]
@Karthiktmv ,

This is likely a source error and not related to KVS. Please try using your source in a GStreamer pipeline set up with other sinks to verify your source is able to reliably stream. Feel free to reopen if you face the issue only while using kvssink, but works with other sinks.

**Comment by Karthiktmv** [comment]
> @Karthiktmv ,
> 
> This is likely a source error and not related to KVS. Please try using your source in a GStreamer pipeline set up with other sinks to verify your source is able to reliably stream. Feel free to reopen if you face the issue only while using kvssink, but works with other sinks.

Hi Disa, I am new to this whole KVS and I am not that technically that aware about how it works, I just want to get it up and working so that i know it's working on my system. Can you kindly help me with the code as well to start the streaming. Also today when i again tried using "gst-device-monitor-1.0" I am getting this for the same intergrated laptop webacm :-

![ss4](https://user-images.githubusercontent.com/111423487/197938732-7e475938-c5bf-46c5-a137-d917f9ebf596.png)


With the same command that I used for streaming earlier from AWS KVS documentation, again I got the internal streaming error.

Here is the output of "gst-inspect-1.0 kvssink" if that will help:-
![ss5](https://user-images.githubusercontent.com/111423487/197940228-456952ba-fefb-4370-8c61-6c9cfac05d27.png)


I know I am a beginner with minimal knowledge just trying to get KVS working, please help me out.

**Comment by sirknightj** [comment]
Hi @Karthiktmv,

What @disa6302 is suggesting is to try a different sink, like [`fakesink`](https://gstreamer.freedesktop.org/documentation/coreelements/fakesink.html) for example. This is to confirm that your v4l2src element is working correctly.

```
gst-launch-1.0 -v v4l2src <properties> ! fakesink
```
